src/EventSubscriber/InitSubscriber.php

```diff
@@ -0,0 +1,145 @@
+   *   Entity type manager.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   Messenger service.
+   * @param \Drupal\password_enhancements\PasswordChecker $password_checker
+   *   Password checker service.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function __construct(AccountProxyInterface $account, DateFormatterInterface $date_formatter, EntityTypeManagerInterface $entity_type_manager, MessengerInterface $messenger, PasswordChecker $password_checker) {
+    $this->account = $account;
+    $this->dateFormatter = $date_formatter;
+    $this->messenger = $messenger;
+    $this->passwordChecker = $password_checker;
+    $this->policyStorage = $entity_type_manager->getStorage('password_enhancements_policy');
+    $this->userStorage = $entity_type_manager->getStorage('user');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    $events[KernelEvents::REQUEST][] = ['checkRequiredPasswordChange', 32];
+    $events[KernelEvents::REQUEST][] = ['passwordChangeNotification', 34];
+    return $events;
+  }
+
+  /**
+   * Checks if the user has to change the password already or not.
+   *
+   * Sets the password change required field to true to force password change if
+   * the user's password expired.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   Response event.
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   *   If the user was failed to save.
+   */
+  public function checkRequiredPasswordChange(GetResponseEvent $event) {
+    if ($event->isMasterRequest()) {
+      $policy = $this->policyStorage->loadByRoleAndPriority($this->account->getRoles());
+      if ($policy !== NULL) {
+        $user = $this->userStorage->load($this->account->id());
+        if (!$user->password_enhancements_password_change_required->value && $this->passwordChecker->isExpired($policy)) {
+          $user->password_enhancements_password_change_required->value = TRUE;
+          $user->save();
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if the user's password is about to expire.
+   *
+   * If the user's password is about to expire it shows an error message if it
+   * is enabled.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   Response event.
+   */
+  public function passwordChangeNotification(GetResponseEvent $event) {
+    if ($event->isMasterRequest() && Url::fromRoute('password_enhancements.password_change')->toString() !== $event->getRequest()->getPathInfo()) {
+      $policy = $this->policyStorage->loadByRoleAndPriority($this->account->getRoles());
+      if ($policy !== NULL && $this->passwordChecker->showWarningMessage($policy)) {
+        $user = $this->userStorage->load($this->account->id());
```

Comment:
How I understand is that account is always an AccountProxy and the user is the loaded User entity. I think renaming these variables like currentUser and loadedCurrentUser could be more confusing because for me the currentUser is the loaded current user Entity.

So I think the $account and  $user is still closer to their class definitions names that harder to confuse with each other.